frontend/src/lib/Api.ts

```diff
@@ -37,6 +41,35 @@ export interface ListRequest {
   sortBy?: string;
 }
 
+type Callback<R> = (err: ServiceError | null, res: R | null) => void;
+type MetadataApiMethod<T, R> = (request: T, callback: Callback<R>) => UnaryResponse;
+type PromiseBasedMetadataApiMethod<T, R> = (request: T) => Promise<{ response: R | null, error: ServiceError | null }>;
+
+/**
+ * Converts a callback based api method to promise based api method.
+ */
+function makePromiseApi<T, R>(apiMethod: MetadataApiMethod<T, R>): PromiseBasedMetadataApiMethod<T, R> {
+  return (request: T) => new Promise((resolve, reject) => {
+    const handler = (error: ServiceError | null, response: R | null) => {
+      // resolve both response and error to keep type information
+      resolve({ response, error });
+    };
+    apiMethod(request, handler);
+  });
+}
+
+const metadataServiceClient = new MetadataStoreServiceClient('');
+
+// TODO: add all other api methods we need here.
+const metadataServicePromiseClient = {
```

Comment:
This is a workaround I made, I think the best approach is to use https://github.com/kubeflow/pipelines/issues/2196.

Migration could be a bad experience when a lot more new code depends on it, so I prefer we do this early. What do you think?

Reply:
Sounds good. I had to switch to grpc-web instead of using @improbable-eng/grpc-web.

Reply:
SG. Is it because the client with promise support is using grpc web client?